File: src/main/java/org/nc/operational/Accountability.java
package org.nc.operational;

import org.nc.knowledge.AccountabilityType;
import org.nc.knowledge.PartyType;

public class Accountability {
    private final AccountabilityType type;
    private final Party commissioner;
    private final Party responsible;
    private final TimePeriod validFor;

    public Accountability(AccountabilityType type, Party commissioner, Party responsible, TimePeriod validFor) {
        this.type = type;
        this.commissioner = commissioner;
        this.responsible = responsible;
        this.validFor = validFor;
        validateConstraints();
    }


    public AccountabilityType getType() {
        return type;
    }

    public Party getCommissioner() {
        return commissioner;
    }

    public Party getResponsible() {
        return responsible;
    }

    public TimePeriod getValidFor() {
        return validFor;
    }

    // Controlla che i tipi di party siano validi per il tipo di accountability
    public void validateConstraints() {
        PartyType commType = commissioner.getType();
        PartyType respType = responsible.getType();

        if (!type.getCommissioners().contains(commType)) {
            throw new IllegalStateException(
                    "Commissioner type '" + commType.getName() +
                            "' is not valid for AccountabilityType '" + type.getName() + "'.");
        }
        if (!type.getResponsibles().contains(respType)) {
            throw new IllegalStateException(
                    "Responsible type '" + respType.getName() +
                            "' is not valid for AccountabilityType '" + type.getName() + "'.");
        }
    }
}

File: python/operational/Accountability.py
from operational.Party import Party
from knowledge.AccountabilityType import AccountabilityType

class Accountability:
    def __init__(self, acc_type: AccountabilityType, commissioner: Party, responsible: Party):
        self.acc_type = acc_type
        self.commissioner = commissioner
        self.responsible = responsible
        self.validate_constraints()

    def validate_constraints(self):
        if not any(t in self.acc_type.valid_commissioners for t in self.commissioner.get_all_types()):
            raise ValueError(f"Commissioner type '{self.commissioner.party_type.name}' is not valid for {self.acc_type.name}")

        if not any(t in self.acc_type.valid_responsibles for t in self.responsible.get_all_types()):
            raise ValueError(f"Responsible type '{self.responsible.party_type.name}' is not valid for {self.acc_type.name}")

    def __repr__(self):
        return f"Accountability({self.acc_type.name}, {self.commissioner.name}, {self.responsible.name})"
